File: proyecto/cppcalc/parser.cpp
#include "parser.h"
#include "calcex.h"
#include "calculator.h"
#include <sstream>

Parser::Parser(istream *in) {
    scan = new Scanner(in);
}

Parser::~Parser() {
    try {
        delete scan;
    } catch (...) { }
}

AST *Parser::parse() {
    return Prog();
}

AST *Parser::Prog() {
    AST *result = Expr();
    Token *t = scan->getToken();

    if (t->getType() != eof) {
        throw ParseError;
    }

    return result;
}

AST *Parser::Expr() {
    return RestExpr(Term());
}

AST *Parser::RestExpr(AST *e) {
    Token *t = scan->getToken();

    if (t->getType() == add)
        return RestExpr(new AddNode(e, Term()));

    if (t->getType() == sub)
        return RestExpr(new SubNode(e, Term()));

    scan->putBackToken();

    return e;
}

AST *Parser::Term() {
    return RestTerm(Storable());
}

AST *Parser::RestTerm(AST *e){
    Token *t = scan->getToken();

    if (t->getType() == modulo)
        return RestTerm(new ModNode(e, Storable()));

    if (t->getType() == times)
        return RestTerm(new TimesNode(e, Storable()));

    if (t->getType() == divide)
        return RestTerm(new DivideNode(e, Storable()));

    scan->putBackToken();

    return e;
}

AST *Parser::Storable() {
    AST *result = Factor();
    Token *t = scan->getToken();
    if(t->getType() == keyword) {
        if (t->getLex() == "S") {
             return new StoreNode(result);
        }else if(t->getLex() == "P") {
            return new MemPlusNode(result);
        }else if(t->getLex() == "M") {
            return new MemMinusNode(result);
        }else{
            throw ParseError;
        }
    }
    scan->putBackToken();
    return result;
}

AST *Parser::Factor() {
    Token *t = scan->getToken();
    if (t->getType() == number){
        istringstream in(t->getLex());
        int val;
        in >> val;
        return new NumNode(val);
    }else if(t->getType() == identifier)
        return Assign(new IdentifierNode(t->getLex()));
    else if(t->getType() == keyword) {
        if (t->getLex() == "R")
            return new RecallNode();
        if (t->getLex() == "C")
            return new ClearNode();
        else{
            throw ParseError;
        }
    }
    else if(t->getType() == lparen)
    {
        AST *result = Expr();
        t = scan->getToken();
        if(t->getType() == rparen)
            return result;
    }
    throw ParseError;
}

AST* Parser::Assign(AST_<string>* e)
{
    Token *t = scan->getToken();
    if(t->getType() == assign)
    {
        AST* result = Expr();
        return new AssignNode(e, result);
    }

    int recall = calc->recallvar(e->evaluate());
    scan->putBackToken();
    return new NumNode(recall);
}

File: proyecto/rubycalc/lib/parser.rb
require 'ast'
require 'scanner'
require 'token'
require 'calcex'

class Parser
  def initialize(istream)
    @scan = Scanner.new(istream)
  end

  def parse
    prog
  end

  private
  def prog
    result = expr
    t = @scan.get_token

    if t.type != :eof
      print "Expected EOF. Found ", t.type, ".\n"
      raise ParseError.new
    end

    result
  end

  def expr
    rest_expr(term)
  end

  def rest_expr(e)
    t = @scan.get_token

    case t.type
      when :add
        return rest_expr(AddNode.new(e, term))
      when :sub
        return rest_expr(SubNode.new(e, term))
    end

    @scan.put_back_token

    e
  end

  def term
    rest_term(storable)
  end

  def rest_term(e)
    t = @scan.get_token

    case t.type
      when :times
        return rest_term(TimesNode.new(e, storable))
      when :divide
        return rest_term(DivideNode.new(e, storable))
      when :modulus
        return rest_term(ModulusNode.new(e, storable))
    end

    @scan.put_back_token

    e
  end

  def storable
    result = factor
    t = @scan.get_token

    if t.type == :keyword
      case t.lex
        when "S"
          return StoreNode.new(result)
        when "P"
          return MemPlusNode.new(result)
        when "M"
          return MemMinusNode.new(result)
        else
          raise ParseError.new
      end
    end

    @scan.put_back_token
    result
  end

  def factor
    t = @scan.get_token
    case t.type
      when :number
        return NumNode.new(t.lex.to_i)
      when :identifier
        return assign(IdentifierNode.new(t.lex))
      when :keyword
        case t.lex
          when "R"
            return RecallNode.new
          when "C"
            return ClearNode.new
          else
            raise ParseError.new
        end
      when :lparen
        result = expr
        t = @scan.get_token
        if t.type == :rparen
          return result
        end
    end
    raise ParseError.new
  end

  def assign(e)
    t = @scan.get_token
    if t.type == :assign
      result = expr
      AssignNode.new(e, result)
    else
      recall = $calc.getvar(e.evaluate)
      @scan.put_back_token
      NumNode.new(recall)
    end
  end
end
